server/auth/oauth2/service.go

```diff
@@ -0,0 +1,147 @@
+package oauth2
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"strings"
+	"time"
+
+	"github.com/argoproj/pkg/rand"
+	"github.com/coreos/go-oidc"
+	log "github.com/sirupsen/logrus"
+	"golang.org/x/oauth2"
+
+	wfv1 "github.com/argoproj/argo/pkg/apis/workflow/v1alpha1"
+)
+
+// https://github.com/dexidp/dex/blob/master/Documentation/using-dex.md
+
+const prefix = "id_token "
+
+type claims struct {
+	Groups []string `json:"groups"`
+}
+
+type Service interface {
+	IsSSO(authorization string) bool
+	Authorize(ctx context.Context, authorization string) (wfv1.User, error)
+	HandleRedirect(writer http.ResponseWriter, request *http.Request)
+	HandleCallback(writer http.ResponseWriter, request *http.Request)
+}
+
+type service struct {
+	config          *oauth2.Config
+	idTokenVerifier *oidc.IDTokenVerifier
+	baseHRef        string
+}
+
+func NewService(issuer, clientID, clientSecret, redirectURL, baseHRef string) (Service, error) {
+	if issuer == "" {
+		return nil, fmt.Errorf("issuer empty")
+	}
+	if clientID == "" {
+		return nil, fmt.Errorf("clientId empty")
+	}
+	if clientSecret == "" {
+		return nil, fmt.Errorf("clientSecret empty")
+	}
+	if redirectURL == "" {
+		return nil, fmt.Errorf("redirectUrl empty")
+	}
+	provider, err := oidc.NewProvider(context.Background(), issuer)
+	if err != nil {
+		return nil, err
+	}
+	config := &oauth2.Config{
+		ClientID:     clientID,
+		ClientSecret: clientSecret,
+		RedirectURL:  redirectURL,
+		Endpoint:     provider.Endpoint(),
+		Scopes:       []string{oidc.ScopeOpenID, "groups"},
+	}
+	idTokenVerifier := provider.Verifier(&oidc.Config{ClientID: clientID})
+	log.WithFields(log.Fields{"redirectURL": config.RedirectURL, "issuer": issuer, "clientId": clientID}).Info("SSO configuration")
+	return &service{config, idTokenVerifier, baseHRef}, nil
+}
+
+const stateCookieName = "oauthState"
+
+func (s *service) HandleRedirect(w http.ResponseWriter, r *http.Request) {
+	state := rand.RandString(10)
+	// TODO - no path?
+	http.SetCookie(w, &http.Cookie{
+		Name:     stateCookieName,
+		Value:    state,
+		Expires:  time.Now().Add(1 * time.Hour),
+		HttpOnly: true,
+		// TODO - lax? not strict?
+		SameSite: http.SameSiteLaxMode,
+	})
+	http.Redirect(w, r, s.config.AuthCodeURL(state), http.StatusFound)
+}
+
+func (s *service) HandleCallback(w http.ResponseWriter, r *http.Request) {
+	ctx := r.Context()
+	state := r.URL.Query().Get("state")
+	cookie, err := r.Cookie(stateCookieName)
+	http.SetCookie(w, &http.Cookie{Name: stateCookieName, MaxAge: 0})
+	if err != nil {
+		w.WriteHeader(400)
+		_, _ = w.Write([]byte(fmt.Sprintf("invalid state: %v", err)))
+		return
+	}
+	if state != cookie.Value {
```

Comment:
The state nonce is supposed to be stored on the server for a short period of time and verified by the **_server_** to be still valid upon return. Adding the state nonce as a cookie adds zero security as the cookie is verified against the query param, which the caller has control over both. 

It's better off removing state verification entirely and leave it as a TODO item. Better would be to use an in-memory time-limited cache of state nonces. However in-memory state nonces precludes more than 1 replica argo-server for SSO login, and then you would have to use something like redis.

Reply:
this was what I was thinking, I did a small amount of research:

https://security.stackexchange.com/questions/140883/is-it-safe-to-store-the-state-parameter-value-in-cookie